pkg/registry/apis/provisioning/jobs/driver.go

```diff
@@ -0,0 +1,211 @@
+
+	// Complete marks a job as completed and moves it to the historic job store.
+	// When in the historic store, there is no more claim on the job.
+	Complete(ctx context.Context, job *provisioning.Job) error
+
+	// Cleanup should be called periodically to clean up abandoned jobs.
+	// An abandoned job is one that has been claimed by a worker, but the worker has not updated the job in a while.
+	Cleanup(ctx context.Context) error
+
+	// InsertNotifications returns a channel that will have a value sent to it when a new job is inserted.
+	// This is used to wake up the job driver when a new job is inserted.
+	InsertNotifications() chan struct{}
+
+	// Update saves the job back to the store.
+	UpdateStatus(ctx context.Context, job *provisioning.Job) (*provisioning.Job, error)
+}
+
+var _ Store = (*persistentStore)(nil)
+
+// jobDriver drives jobs to completion and manages the job queue.
+// There may be multiple jobDrivers running in parallel.
+// The jobDriver deals with cleaning up upon death and ensuring that jobs remain claimable.
+type jobDriver struct {
+	// Timeout for processing a job. This should be the same or less than a claim expiry.
+	timeout time.Duration
+	// CleanupInterval is the time between cleanup runs.
+	cleanupInterval time.Duration
+	// JobInterval is the time between job ticks. This should be relatively low.
+	jobInterval time.Duration
+
+	// Store is the job storage backend.
+	store Store
+	// RepoGetter lets us access repositories to pass to the worker.
+	repoGetter RepoGetter
+
+	// Workers process the job.
+	// Only the first worker who supports the job will process it; the rest are ignored.
+	workers []Worker
+}
+
+func NewJobDriver(
+	timeout, cleanupInterval, jobInterval time.Duration,
+	store Store,
+	repoGetter RepoGetter,
+	workers ...Worker,
+) *jobDriver {
+	return &jobDriver{
+		timeout:         timeout,
+		cleanupInterval: cleanupInterval,
+		jobInterval:     jobInterval,
+		store:           store,
+		repoGetter:      repoGetter,
+		workers:         workers,
+	}
+}
+
+// Run drives jobs to completion. This is a blocking function.
+// It will run until the context is canceled.
+// This is a thread-safe function; it may be called from multiple goroutines.
+func (d *jobDriver) Run(ctx context.Context) {
+	cleanupTicker := time.NewTicker(d.cleanupInterval)
+	defer cleanupTicker.Stop()
+
+	jobTicker := time.NewTicker(d.jobInterval)
+	defer jobTicker.Stop()
+
+	logger := logging.FromContext(ctx).With("logger", "job-driver")
+	ctx = logging.Context(ctx, logger)
+	ctx, _, err := identity.WithProvisioningIdentity(ctx, "*") // "*" grants us access to all namespaces.
+	if err != nil {
+		logger.Error("failed to grant provisioning identity; this will panic!", "error", err)
+		panic("unreachable?: failed to grant provisioning identity: " + err.Error())
+	}
+
+	// Drive without waiting on startup.
+	d.startDriving(ctx)
+
+	for {
+		select {
+		case <-cleanupTicker.C:
```

Comment:
can you explain why you decided to implement the cleanup ticker outside the store? I would have expected the component accepting the claims controlling that as well.

Reply:
I want to keep the store as a storage layer and nothing more: it only interacts with the underlying storage. The driver does all the asynchronous processing for the storage.

I'm not too fussed about what owns what, though I would prefer having less logic on each by splitting them.

Reply:
yeah, but if you look at how the Claim works, it has more logic that a simply database call. It behaves more like claming service for which the clean up logic would make it more self-contained.